bot/cogs/doc.py

```diff
@@ -0,0 +1,542 @@
+import asyncio
+import functools
+import logging
+import random
+import re
+from collections import OrderedDict
+from ssl import CertificateError
+from typing import Dict, List, Optional, Tuple
+
+import discord
+from aiohttp import ClientConnectorError
+from bs4 import BeautifulSoup
+from discord.ext import commands
+from markdownify import MarkdownConverter
+from requests import ConnectionError
+from sphinx.ext import intersphinx
+
+from bot.constants import ERROR_REPLIES, Keys, Roles, URLs
+from bot.decorators import with_role
+
+
+log = logging.getLogger(__name__)
+logging.getLogger('urllib3').setLevel(logging.WARNING)
+
+
+UNWANTED_SIGNATURE_SYMBOLS = ('[source]', '¶')
+WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)(\s+)")
+
+
+def async_cache(max_size=128, arg_offset=0):
+    """
+    LRU cache implementation for coroutines.
+
+    :param max_size:
+    Specifies the maximum size the cache should have.
+    Once it exceeds the maximum size, keys are deleted in FIFO order.
+    :param arg_offset:
+    The offset that should be applied to the coroutine's arguments
+    when creating the cache key. Defaults to `0`.
+    """
+
+    # Assign the cache to the function itself so we can clear it from outside.
+    async_cache.cache = OrderedDict()
+
+    def decorator(function):
+        @functools.wraps(function)
+        async def wrapper(*args):
+            key = ':'.join(args[arg_offset:])
+
+            value = async_cache.cache.get(key)
+            if value is None:
+                if len(async_cache.cache) > max_size:
+                    async_cache.cache.popitem(last=False)
+
+                async_cache.cache[key] = await function(*args)
+            return async_cache.cache[key]
+        return wrapper
+    return decorator
+
+
+class DocMarkdownConverter(MarkdownConverter):
+    def convert_code(self, el, text):
+        """Undo `markdownify`s underscore escaping."""
+
+        return f"`{text}`".replace('\\', '')
+
+    def convert_pre(self, el, text):
+        """Wrap any codeblocks in `py` for syntax highlighting."""
+
+        code = ''.join(el.strings)
+        return f"```py\n{code}```"
+
+
+def markdownify(html):
+    return DocMarkdownConverter(bullets='•').convert(html)
+
+
+class DummyObject(object):
+    """
+    A dummy object which supports assigning anything,
+    which the builtin `object()` does not support normally.
+    """
+
+
+class SphinxConfiguration:
+    """Dummy configuration for use with intersphinx."""
+
+    config = DummyObject()
+    config.intersphinx_timeout = 3
+    config.tls_verify = True
+
+
+class ValidPythonIdentifier(commands.Converter):
```

Comment:
We do have a converters file, converters should probably be in there.

Reply:
That would make the converters module a bit messy, IMO. For example, the `DocumentationBaseURL` and `InventoryURL` converters makes use of:
- `aiohttp.ClientConnectorError`
- `ssl.CertificateError`
- `intersphinx`
- `bot.cogs.doc.SphinxConfiguration`

I can move it there if it's really wanted, but it would make more sense to keep it in the cog IMO - at least the `InventoryURL` converter. I can see `DocumentationBaseURL` (renamed to something like `ValidURL`) and `ValidPythonIdentifier` having some general use, but I can't really see `InventoryURL` being used in another cog.

Reply:
Have you seen the contents of that file? It has a bunch of snake cog stuff in it.

It's really just a maintainability thing. But if it would be messier to move it, then I guess you could just not.

Reply:
I have, yes. But I could move it. I'm okay with both, so you tell me 🤷‍♂️

Reply:
Do you feel like it's more effort than it's worth?

Reply:
No, will do